sdk/cosmosdb/azure_data_cosmos/src/clients/database_client.rs

```diff
@@ -0,0 +1,51 @@
+use crate::authorization_policy::ResourceType;
+use crate::models::DatabaseProperties;
+use crate::{CosmosClient, ReadDatabaseOptions};
+use azure_core::{Context, Request};
+use url::Url;
+
+pub trait DatabaseClientMethods {
+    fn read(
+        &self,
+        options: Option<ReadDatabaseOptions>,
+    ) -> impl std::future::Future<
+        Output = azure_core::Result<azure_core::Response<DatabaseProperties>>,
+    > + Send;
```

Comment:
Our guidelines recommend using `async fn` in traits, but that runs counter to the guidance in the `async_fn_in_trait` warning:

```
warning: use of `async fn` in public traits is discouraged as auto trait bounds cannot be specified
 --> sdk/cosmosdb/azure_data_cosmos/src/clients/database_client.rs:8:5
  |
8 |     async fn read(
  |     ^^^^^
  |
  = note: you can suppress this lint if you plan to use the trait only in your own code, or do not care about auto traits like `Send` on the `Future`
  = note: `#[warn(async_fn_in_trait)]` on by default
help: you can alternatively desugar to a normal `fn` that returns `impl Future` and add any desired bounds such as `Send`, but these cannot be relaxed without a breaking API change
  |
8 ~     fn read(
9 |         &self,
10|         options: Option<ReadDatabaseOptions>,
11~     ) -> impl std::future::Future<Output = azure_core::Result<azure_core::Response<DatabaseProperties>>> + Send;
  |

```

Reply:
I don't think we _need_ this to be `Send` or `Sync`, but the point made by the warning is still valid. By using `async fn`, it make not be clear that adding one of those auto-trait bounds would be breaking.

Reply:
It's also important to note that if we desugar the trait method as recommended by the warning, the `impl` **can still be an `async fn`** so we still retain the ability to use `async`/`await` in implementations of the trait.

Reply:
In fact, this warning is causing a build failure since we fail on all lints in CI at the moment.

Reply:
#1796

Reply:
I left a fairly detailed comment there with some of my understanding of the problem and proposal for how to move forward. For now, I'll do as suggested and use `async fn` and `expect` the lint to suppress it.

Reply:
I'll `allow` it actually. The `expect` attribute was only just stabilized in `1.81` and appears it's not possible to condition attributes on compiler version in Stable rust either.